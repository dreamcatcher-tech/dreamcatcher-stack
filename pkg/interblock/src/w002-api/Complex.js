import assert from 'assert-fast'
export default class Complex {
  #parent
  static create({ state, binary, network, actions, isLoading, wd, tree }) {
    return new Complex({ state, binary, network, actions, isLoading, wd, tree })
  }
  constructor({ state, binary, network, actions, isLoading, wd, tree }) {
    this.state = state || {}
    this.binary = binary
    this.network = network || []
    this.actions = actions || {}
    this.isLoading = isLoading || false
    this.wd = wd || '/'
    this.tree = tree
    if (tree === undefined) {
      this.tree = Complex.create({ ...this, tree: false })
    }
  }
  clone() {
    const next = new Complex(this)
    next.#parent = this.#parent
    return next
  }
  addAction(actions) {
    const next = this.clone()
    next.actions = { ...next.actions, ...actions }
    return next
  }
  setWd(wd) {
    assert.strictEqual(typeof wd, 'string')
    assert(wd.startsWith('/'))
    const next = this.clone()
    next.wd = wd
    return next
  }
  child(path) {
    assert.strictEqual(typeof path, 'string')
    assert(!path.startsWith('/'))
    assert(!path.startsWith('./'))
    assert(!path.startsWith('..'))
    const child = this.network.find(({ path: p }) => p === path)
    if (!child) {
      throw new Error(`child not found: ${path}`)
    }
    const { tree } = this
    const next = Complex.create({ ...child, tree })
    next.#parent = this
    return next
  }
  hasChild(path) {
    if (!path) {
      return false
    }
    assert.strictEqual(typeof path, 'string')
    assert(!path.startsWith('/'))
    assert(!path.startsWith('./'))
    assert(!path.startsWith('..'))
    return this.network.some(({ path: p }) => p === path)
  }
  rm(path) {
    assert(this.hasChild(path))
    const next = this.clone()
    next.network = next.network.filter(({ path: p }) => p !== path)
    return next
  }
  setNetwork(network) {
    assert(Array.isArray(network))
    const next = this.clone()
    next.network = network
    return next
  }
  setState(state) {
    assert.strictEqual(typeof state, 'object')
    const next = this.clone()
    next.state = state
    return next
  }
  parent() {
    return this.#parent
  }
}
