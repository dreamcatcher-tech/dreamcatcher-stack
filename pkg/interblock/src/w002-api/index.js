import assert from 'assert-fast'

export const interchain = (type, payload, to) => {
  const { interchain } = globalThis[Symbol.for('interblock:api:hook')]
  assert(interchain, `global hook detached`)
  return interchain(type, payload, to)
}
