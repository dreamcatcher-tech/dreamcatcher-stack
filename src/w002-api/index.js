const api = require('./src/api')
const hooks = require('./src/hooks')
const queries = require('./src/queries')
module.exports = { ...api, ...hooks, ...queries }
