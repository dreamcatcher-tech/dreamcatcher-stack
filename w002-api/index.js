const api = require('./src/api')
const promises = require('./src/promises')
module.exports = { ...api, ...promises }
