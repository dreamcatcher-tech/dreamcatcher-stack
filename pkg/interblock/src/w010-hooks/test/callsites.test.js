import { assert } from 'chai/index.mjs'
import { interchain } from '../../w002-api'
import { wrapReduce } from '..'
import Debug from 'debug'
import {
  AsyncRequest,
  AsyncTrail,
  Reply,
  Request,
  RequestId,
  RxRequest,
  RxReply,
} from '../../w008-ipld'
const debug = Debug('interblock:tests:hooks')
Debug.enable()
const request = Request.create('TEST')
const requestId = RequestId.createCI()
const rxRequest = RxRequest.create(request, requestId)
const init = AsyncTrail.create(rxRequest)
/**
 * Test callsites for resolve and for multiple promises
 * Test using prepopulated accumulators
 *
 * Then test the engine using prepop accumulators for specific actions
 * to avoid code relooping and making it hard to trace
 *
 * test the engine incrementally with each new system function
 *
 *
 */
describe('callsites', () => {
  describe('basics', () => {
    const nested =
      (id, depth = 0) =>
      async () => {
        if (depth === 0) {
          return { id }
        } else {
          interchain(`id: ${id} depth: ${depth}`)
          return nested(id, depth - 1)()
        }
      }

    test('nested hooks awaited', async () => {
      const id = 57
      const depth = 1000
      const reducer = nested(id, depth)
      const trail = await wrapReduce(init, reducer)

      assert(trail.isPending())
      assert(!trail.reply)
      assert(Array.isArray(trail.txs))
      assert.strictEqual(trail.txs.length, depth)
    })
    test('void return', async () => {
      const voidReturn = () => {
        return
      }
      const trail = await wrapReduce(init, voidReturn)
      assert(!trail.isPending())
      assert(trail.reply.isResolve())
      assert.deepEqual(trail.reply.payload, {})
      assert.strictEqual(trail.txs.length, 0)
    })
    test('plain reply', async () => {
      const payload = { plain: true }
      const plain = () => payload
      const trail = await wrapReduce(init, plain)
      assert(!trail.isPending())
      assert(trail.reply.isResolve())
      assert.strictEqual(trail.reply.payload, payload)
      assert.strictEqual(trail.txs.length, 0)
    })
    test.todo('nested calls to wrapReduce')
    test('nested parallel hooks do not collide', async () => {
      const inits = Array(4).fill(true)
      const nestedDepth = 4
      const awaits = inits.map((_, index) => {
        const reducer = nested(index, nestedDepth + index)
        return wrapReduce(init, reducer)
      })
      const results = await Promise.all(awaits)
      expect(results).toMatchSnapshot()
    })
    test('reduction must be an object', async () => {
      const msg = 'Must return either undefined, or an object'
      const e1 = await wrapReduce(init, () => () => 'arrow fn')
      assert.strictEqual(e1.getError().message, msg)
      const e2 = await wrapReduce(init, () => true)
      assert.strictEqual(e2.getError().message, msg)
      const e3 = await wrapReduce(init, () => 'string')
      assert.strictEqual(e3.getError().message, msg)
      const e4 = await wrapReduce(init, () => 5)
      assert.strictEqual(e4.getError().message, msg)
    })
    test('duplicate requests permitted in same call', async () => {
      const double = async () => {
        interchain('twin')
        interchain('twin')
        // TODO supply a response, and verify the second request gets a different response
        return {}
      }
      const trail = await wrapReduce(init, double)
      const { txs } = trail
      assert.strictEqual(txs.length, 2)
      assert(trail.isPending())
    })
    test('resolved result', async () => {
      const resolved = async () => {
        await Promise.resolve()
        return { resolved: true }
      }
      const trail = await wrapReduce(init, resolved)
      const { reply, txs } = trail
      assert.strictEqual(txs.length, 0)
      assert(reply.isResolve())
      assert(!trail.isPending())
    })
    test('reject', async () => {
      const rejector = async () => {
        throw new Error('rejected')
      }
      const trail = await wrapReduce(init, rejector)
      assert.strictEqual(trail.getError().message, 'rejected')
      assert.strictEqual(trail.isPending(), false)
    })
    test('timeout exceeded', async () => {
      const slowest = () => {
        return new Promise(() => Infinity)
      }
      const trail = await wrapReduce(init, slowest)
      assert.strictEqual(trail.getError().message, 'timeout exceeded: 100 ms')
    })
    test.todo('duplicate requests permitted in different calls')
  })
  describe('with asyncs', () => {
    test('single await', async () => {
      let interchainResult
      const single = async () => {
        interchainResult = await interchain('single')
      }
      let trail = await wrapReduce(init, single)
      assert.strictEqual(interchainResult, undefined)
      assert(trail.isPending())
      assert.strictEqual(trail.txs.length, 1)
      let [pendingRequest] = trail.txs
      assert(pendingRequest instanceof AsyncRequest)
      expect(pendingRequest).toMatchSnapshot()
      pendingRequest = pendingRequest.setId(requestId)
      trail = trail.updateTxs([pendingRequest])
      const reply = Reply.create('@@RESOLVE', { single: true })
      const rxReply = RxReply.create(reply, requestId)
      trail = trail.settleTx(rxReply)
      trail = await wrapReduce(trail, single)
      assert.strictEqual(trail.txs.length, 0)
      assert.strictEqual(interchainResult.single, true)
      assert(trail.reply.isResolve())
      assert.deepEqual(trail.reply.payload, {})
    })
    test('rejection', async () => {
      let interchainResult
      const reject = async () => {
        try {
          interchainResult = await interchain('single')
        } catch (error) {
          interchainResult = error
          throw error
        }
      }
      let trail = await wrapReduce(init, reject)
      assert.strictEqual(interchainResult, undefined)
      let [pendingRequest] = trail.txs
      pendingRequest = pendingRequest.setId(requestId)
      trail = trail.updateTxs([pendingRequest])
      const rejection = Reply.create('@@REJECT', new Error('test rejection'))
      const rxReply = RxReply.create(rejection, requestId)
      trail = trail.settleTx(rxReply)
      trail = await wrapReduce(trail, reject)
      const error = trail.getError()
      assert.strictEqual(error.message, 'test rejection')
      assert(interchainResult instanceof Error)
    })
    test.todo('throw on incomplete accumulator')
  })
})
