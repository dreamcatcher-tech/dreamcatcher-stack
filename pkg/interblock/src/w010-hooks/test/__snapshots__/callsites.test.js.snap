// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callsites basics nested parallel hooks do not collide 1`] = `
Array [
  Reduction {
    "txs": Array [
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 0 depth: 4",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 0 depth: 3",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 0 depth: 2",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 0 depth: 1",
        },
        "to": ".",
      },
    ],
  },
  Reduction {
    "txs": Array [
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 1 depth: 5",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 1 depth: 4",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 1 depth: 3",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 1 depth: 2",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 1 depth: 1",
        },
        "to": ".",
      },
    ],
  },
  Reduction {
    "txs": Array [
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 2 depth: 6",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 2 depth: 5",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 2 depth: 4",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 2 depth: 3",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 2 depth: 2",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 2 depth: 1",
        },
        "to": ".",
      },
    ],
  },
  Reduction {
    "txs": Array [
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 3 depth: 7",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 3 depth: 6",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 3 depth: 5",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 3 depth: 4",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 3 depth: 3",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 3 depth: 2",
        },
        "to": ".",
      },
      PendingRequest {
        "request": Request {
          "payload": Object {},
          "type": "id: 3 depth: 1",
        },
        "to": ".",
      },
    ],
  },
]
`;

exports[`callsites with accumulator single await 1`] = `
PendingRequest {
  "request": Request {
    "payload": Object {},
    "type": "single",
  },
  "to": ".",
}
`;
